libraries/extractor/src/main/java/androidx/media3/extractor/DtsUtil.java

```diff
@@ -223,8 +357,262 @@ public static int getDtsFrameSize(byte[] data) {
+          frameBits.skipBits(4); // nuAssetTypeDescriptor
+        }
+        if (frameBits.readBits(1) == 1) { // bLanguageDescrPresent
+          frameBits.skipBits(24); // LanguageDescriptor
+        }
+        if (frameBits.readBits(1) == 1) { // bInfoTextPresent
+          int nuInfoTextByteSize = frameBits.readBits(10) + 1;
+          frameBits.skipBits(nuInfoTextByteSize * 8); // InfoTextString
+        }
+        frameBits.skipBits(5); // nuBitResolution
+
+        sampleRate = SAMPLE_RATE_EXTSS[frameBits.readBits(4)];
+        channelCount = frameBits.readBits(8) + 1;
+      }
+    }
+
+    // Number of audio sample in a compressed DTS frame
+    int sampleCount = nExSSFrameDurationCode * (sampleRate / REF_CLOCK_TABLE[ nRefClockCode]);
+    return new DtsFormatInfo(
+        MimeTypes.AUDIO_DTS_EXPRESS,
+        channelCount,
+        sampleRate,
+        extFrameSize,
+        sampleCount,
+        0);
+  }
+
+  /**
+   * Returns the size of frame header in a DTS-HD frame by parsing a few bytes of data
+   * (minimum 8 bytes) from the input bitstream(containing only Extension Sub-stream).
+   *
+   * @param frame The initial 8 bytes(minimum) of a DTS-HD frame(having only Extension Sub-stream).
+   * @return Size of the DTS-HD frame header in bytes.
+   */
+  public static int parseDtsHdHeaderSize(byte[] frame) {
+    ParsableBitArray frameBits = getNormalizedFrameHeader(frame);
+    int nuBits4Header;
+    frameBits.skipBits( 32 + 8 + 2); // SYNC, UserDefinedBits, nExtSSIndex
+    // Unpack the num of bits to be used to read header size
+    int headerSizeType = frameBits.readBits(1);
+    if (headerSizeType == 0){
+      nuBits4Header = 8;
+    }
+    else{
+      nuBits4Header = 12;
+    }
+    // Unpack the substream header size
+    return frameBits.readBits(nuBits4Header) + 1;
+  }
+
+  /**
+   * Returns the DTS format given {@code data} containing the DTS-UHD(Profile 2) frame according to
+   * ETSI TS 103 491 V1.2.1 (2019-05), Section 6.4.3.
+   *
+   * @param frame The DTS-UHD frame to parse.
+   * @return The DTS format parsed from data in the header.
+   */
+  public static DtsFormatInfo parseDtsUhdFormat(byte[] frame) throws ParserException {
+    ParsableBitArray frameBits = getNormalizedFrameHeader(frame);
+    int sampleRate = 0;
+    int sampleCount = 0;
+    boolean bSyncFrameFlag = false;
+
+    int nSyncWord = frameBits.readBits(32);
+    if (nSyncWord == SYNC_VALUE_UHD_FTOC_SYNC_BE) {
+      bSyncFrameFlag = true;
+    }
+    int[] ucTable1 = new int[] { 5, 8, 10, 12 };
+    int nFTOCPayloadinBytes = extractVarLenBitFields(frameBits, ucTable1, true) + 1;
+
+    if (bSyncFrameFlag) {
+      // fullChannelBasedMixFlag, ETSI TS 103 491 V1.2.1, Section 6.4.6.1
+      if(frameBits.readBits(1) != 1) {
+        throw ParserException.createForMalformedContainer(
+            /* message= */ "Only supports full channel mask-based audio presentation",
+            /* cause= */ null);
+      }
+
```

Comment:
should there be a CRC check here? 6.4.6.2

Reply:
The CRC check is not valid for fully channel-based mix(m_bFullChannelBasedMixFlag = TRUE).
Currently, we encode such streams only.

Reply:
Hey @rahulnmohan, I'm reviewing this PR internally - I thought it would be quicker to come back you with some unresolved questions.

I'm not sure I understand how what you've said here reconciles with the spec that says:
> When `m_bFullChannelBasedMixFlag = TRUE`, the FTOC CRC16 word is transmitted in the stream only in the sync frames. Consequently for this case the CRC check is performed only in sync frames.

It seems that here we **are** in a sync frame (we're inside an `if (syncFrameFlag)` stanza), so according to the spec we should be doing a CRC check here. Can you help me understand why you're saying it would be invalid to do so?

Reply:
Hi @icbaker,
Thank you very much for reviewing the code.
I have added the CRC check for the FTOC sync frames. Could you check the code updates?